java-checks/src/main/java/org/sonar/java/checks/sustainability/AndroidExactAlarmCheck.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.sustainability;
+
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatchersBuilder;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S6891")
+public class AndroidExactAlarmCheck extends AbstractMethodDetection {
+
+  private static final String SET_EXACT_NAME = "setExact";
+
+  private static final String SET_EXACT_AND_ALLOW_WHILE_IDLE_NAME = "setExactAndAllowWhileIdle";
+
+  private static final String SET_WINDOW_NAME = "setWindow";
+
+  private static final long SUGGESTED_MIN_LENGTH_MILLIS = 10L * 60L * 1000L;
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.create()
+      .ofSubTypes("android.app.AlarmManager")
+      .names(SET_EXACT_NAME, SET_EXACT_AND_ALLOW_WHILE_IDLE_NAME, SET_WINDOW_NAME)
+      .withAnyParameters()
+      .build();
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree tree) {
+    switch (MethodMatchersBuilder.getIdentifier(tree).name()) {
+      case SET_EXACT_NAME:
+      case SET_EXACT_AND_ALLOW_WHILE_IDLE_NAME:
+        onSetExactInvocation(tree);
+        break;
+      case SET_WINDOW_NAME:
+        onSetWindowInvocation(tree);
+        break;
+      default:
+        throw new IllegalArgumentException();
```

Comment:
How about adding an error message here? Alternatively logging an error instead of throwing an exception?

Reply:
I don't think it's needed: this is just to make the `switch` clause exhaustive. It will never happen because the `MethodMatchers` have already made sure it's one of those three functions. If it still happens, this is a programming mistake (e.g., one more matching function added to the `MethodMatchers`, but not to this switch statement). I think, for this purpose, no logging and just an exception (without a message) to disrupt the control flow is OK.